lib/tree.rb

```diff
@@ -16,49 +16,158 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) for call stack
+  def add(key, value = nil)
```

Comment:
👍 The time complexity is O(log n) if the tree is balanced and O(n) if it isn't.  The space complexity is the same.